### PR TITLE
Convert Addon Proposal to Application Proposal

### LINKED ISSUE #8304: Convert Addon-Proposal to Application-Proposal
In a separate meeting, we decided to move away from managing DefaultAddons due to the following reasons:

- Default Addons are distinct from Custom Addons and often require other components outside of the user cluster (e.g. kube-metrics)
- We decided that decoupling Default Addons from KKP versions is very difficult to support. This is because it allows the usage of all combinations of KKP and Addon versions, which are hard to test for us in advance. As a result we could run into a variety of unexpected errors

**User Story**

As a result we have to adapt our original proposal to adapt to these new changes. Additionally in order to avoid confusion with the existing Addon-mechanis, it was decided to rename the new approach into "Applications"

### PR DESCRIPTION
**Pull request by SimonTheLeg**
Signed-off-by: Simon Bein <simontheleg@gmail.com>

**What this PR does / why we need it**
Context for the change: After the inital merge in #8125, there have been concerns regarding tackling both Default and Custom Addons. In the end, we have decided to simplify the proposal and only focus on components a user should control in their cluster, not mandatory components provided by Kubermatic.
We decided to move a way from this due to the following factors:
  - Default Addons are distinct from Custom Addons and often require other components outside of the user cluster (e.g. kube-metrics)
  - We decided that decoupling Default Addons from KKP versions is very difficult to support. This is because it allows the usage of all combinations of KKP and Addon versions, which are hard to test for us in advance. As a result we 

**Side Notes**
The file is currently still called `improve-addons.md`. This is so git displays a proper diff, which makes it easier for reviewing. Shortly before merging, I am going to rename the file.

**Which issue(s) this PR fixes** 
Pre-req for #7974, solves #8221
fix #8304

**Does this PR introduce a user-facing change?**:

```release-note
None
```

### CODE REVIEW

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]:
```suggestion
Note on CRDs in general: The ApplicationDefinition strongly benefits from [OpenAPIv3's `oneOf` functionality](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/), which is supported by Kubernetes. For example in the yaml above, you could say a source must be `oneOf` `helm` or `git`. The cool thing about `oneOf` is that `helm` and `git` can have totally have different fields from each other, but still would be statically validated. However, syntax like this is [not yet supported in kubebuilder](https://github.com/kubernetes-sigs/controller-tools/issues/461), which we use to generate the CRDs. Nonetheless, we would still propose to use kubebuilder for now as the advantages of an automated generation outweigh the downside of not being able to do oneOf (yet).
```

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]:
```suggestion
- adding an endpoint to install an application in a user cluster. It is important to note that it only creates the CRs which Application should be installed in which cluster, but does not do the installation of the Application itself (this task will be done by the ApplicationInstallationController).
```

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]: In fact, it may have an impact because the cluster template is stored in `clustertemplate` and cluster are created from `clustertemplateinstances` CR.  So I have to check if we may encounter the same kind of problem regarding the size.

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]:
```suggestion
    name: prometheus-node-exporter
```
to be coherent with the example of  `ApplicationDefinition` CR

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]:
```suggestion
The functionality of the ApplicationInstallationController depends on the selected `source` and `method` in the ApplicationDefinition CR. To start we propose the following sources and methods:
```

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]: i would move this part, next to ApplicationDefinition CR

File: docs/proposals/improve-addons.md
Comment (SimonTheLeg) [reply docs/proposals/improve-addons.md]: updated section with a new idea @vgramer and I developed

File: docs/proposals/improve-addons.md
Comment (SimonTheLeg) [reply docs/proposals/improve-addons.md]: I think this would break this paragraph a bit. Because in the bullets we often describe what the ApplicationInstallationController currently is doing for the given source.

Comment (vgramer) [reply docs/proposals/improve-addons.md]: fair enough

**Comment by SimonTheLeg** [comment]
/hold

Add hold label, so this PR does not get auto-merged after the first person has approved it.

**Comment by rastislavs** [comment]
/lgtm

Minor comment on the diagram "Cluster Administrator Creates A new Cluster with Addons", Addons / Addon Controller were not renamed, not sure if intentionally (as TBD is present in the title) or not.

**Comment by SimonTheLeg** [comment]
@rastislavs ah good point! I have updated the diagram.

File: docs/proposals/improve-addons.md
Comment (kron4eg) [inline docs/proposals/improve-addons.md]:
```suggestion
  namespace: cluster-47s2ddlfgj
```

**Review by kron4eg — COMMENTED** [review]
/approve

micro nit in the comment

**Comment by kron4eg** [comment]
I'm playing in the mind with the idea of custom resources definitions presented here would consist a dedicated group (for example, apps.kubermatic.k8c.io or any other).

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]:
```suggestion
apiVersion: apps.kubermatic.k8c.io/v1
```

File: docs/proposals/improve-addons.md
Comment (vgramer) [inline docs/proposals/improve-addons.md]:
```suggestion
apiVersion: apps.kubermatic.k8c.io/v1
```

**Comment by vgramer** [comment]
> I'm playing in the mind with the idea of custom resources definitions presented here would consist a dedicated group (for example, apps.kubermatic.k8c.io or any other).

make sense to me.

**Comment by SimonTheLeg** [comment]
@vgramer is faster than the wind. I was about to commit the `apps` addition :D

File: docs/proposals/improve-addons.md
Comment (mfranczy) [inline docs/proposals/improve-addons.md]: I would kill this idea, we are not building CI/CD, that should be handled by customers.

File: docs/proposals/improve-addons.md
Comment (mfranczy) [inline docs/proposals/improve-addons.md]: I think adding a new endpoint for creating application CR is better. You would not have to refactor the `CreateEndpoint`. I think UI should handle requests for creating clusters and later requests for applications when the cluster is ready. It could be done over some mechanims like promises. However, we have to ask about that somebody from the UI team. @maciaszczykm what do you think?

EDIT: although giving this a second thought, it might not be a bad idea to extend cluster. I don't have very strong opinion here.

File: docs/proposals/improve-addons.md
Comment (mfranczy) [inline docs/proposals/improve-addons.md]: As I understand the installation controller would be deployed on the seed in the kubermatic namespace and watch all user clusters. I think we should change that. Instead of having relation one to many in that case it's better to have relation one to one.

It could be that we deploy application installation controller in the cluster namespace and it deploys ApplicationInstallation CR in the user cluster (similar like Machine Controller works and creates machines objects on the user cluster).

@moadqassem @kron4eg any opinion on this?

File: docs/proposals/improve-addons.md
Comment (mfranczy) [reply docs/proposals/improve-addons.md]: btw please change `AddonsRef` to `ApplicationRef`

File: docs/proposals/improve-addons.md
Comment (mfranczy) [inline docs/proposals/improve-addons.md]: IMO should be deployed on the user cluster not on the seed.

File: docs/proposals/improve-addons.md
Comment (mfranczy) [inline docs/proposals/improve-addons.md]: I would change:
> Therefore dependency management outside an Application is currently out of scope
to
>  Therefore dependency management outside a supported source type is out of scope.

Or something similar to this, what I mean is that helm support dependencies. The sentence indicates that application has to bundle all dependencies (which is not true in case of helm).

File: docs/proposals/improve-addons.md
Comment (mfranczy) [reply docs/proposals/improve-addons.md]: My reasoning here is that `ApplicationInstallation`  is user cluster specific those should be deployed there.

File: docs/proposals/improve-addons.md
Comment (kron4eg) [reply docs/proposals/improve-addons.md]: It makes sense if we want to expose ApplicationInstallation CR to the end user, in the user-cluster API. A-la MachineDeployment. But the question is: is it the case?

Comment (mfranczy) [reply docs/proposals/improve-addons.md]: I think it is, in my opinion this should be exposed to the end user (user cluster admin). @SimonTheLeg @vgramer what's your point of view on this?

Comment (mfranczy) [reply docs/proposals/improve-addons.md]: and @shaase-ctrl ^

File: docs/proposals/improve-addons.md
Comment (maciaszczykm) [reply docs/proposals/improve-addons.md]: For initial addons/applications (ones created with the cluster; from wizard) I'd say it would be the best to keep them as part of the cluster object (perhaps in the annotation like initial node deployment) and then create addons/application CR? For the existing clusters I'd use CRs only.

cc @kubermatic/sig-ui @kubermatic/sig-ui for more opinions on this

File: docs/proposals/improve-addons.md
Comment (shaase-ctrl) [reply docs/proposals/improve-addons.md]: Sorry, I did not make it in time over the weekend. 
We need to have a user cluster admin be able to install apps for his cluster. KKP admins that need to configure everything per cluster is contrary to the self-service platform we built.

Comment (kron4eg) [reply docs/proposals/improve-addons.md]: I don't think it's possible to expose it, as it's defined as

> contains a mapping of a specific Application and cluster where to install

Comment (vgramer) [reply docs/proposals/improve-addons.md]: With the current design, KKP admin defined possible applications to install by creating an `ApplicationDefinition` CR in the master cluster. The `ApplicationDistributor` controller distributes these CRs to all seeds. 

the `ApplicationInstallation` CR  contains a reference to the `ApplicationDefinition` CR and the version of the application. this reference is used to know how to install the application.

if we move the `ApplicationInstallation` to the user-cluster and the `ApplicationInstallationController` into the cluster-namespace. That means we have to either:

1.  have a client to the seed cluster in the `ApplicationInstallationController` to be able to retrieve to the corresponding `ApplicationDefinition` CR. 
2. have a controller (ApplicationDistributor controller ?) that replicates `ApplicationDefinition` CR into each user cluster. So the `ApplicationInstallationController` can retrieve it.


Solution 1 seems pretty messy to me. especially if we consider deletion of an `ApplicationDefinition` CR  by a KKP admin. what happens to applications already installed.  

With solution 2, the deletion can be handled by an ownerRef set to `ApplicationInstalltion` CR inside the user-cluster.  When a KKP admin deletes an `ApplicationDefinition` CR, the deletion is eventually propagated to user-cluster and applications are removed.

Comment (SimonTheLeg) [reply docs/proposals/improve-addons.md]: In addition to @vgramer comments, maybe a short background and two follow-up questions from my side

Background:
We had two initial ideas behind placing the ApplicationInstallationCRs inside the seed:
1. It would be tough for cluster admins to create invalid ApplicationInstallationCRs, as they can only use the UI to create them. This was born out of the idea that the majority of Application users would use the UI. Which sounds plausible to me, because if someone is more proficient in k8s and knows how to use helm, kustomized, etc., they would not really need the KKP Application feature. => now I think we can drop this issue, as we extend our ValidatingWebhook to ensure that no invalid ApplicationInstallationCRs can be created/edited (I think we just need to add that users cannot change the status field, otherwise Application state could get messy). @shaase-ctrl from a product-mgmt perspective: should the Cluster admin be able to edit ApplicationInstallations using the k8s api or would it be sufficient if they can do it via the UI?
2. Reduce the number of cross-cluster calls. We need to be aware of the fact that moving the ApplicationInstallationCRs into the user clusters will result in more cross-cluster calls. This is due to the fact that we need to keep ApplicationDefinitions in the seed cluster (otherwise cluster admins could just edit the definitions and make up their own Applications). This in turn means that for almost every interaction with the ApplicationInstallation we need to call the seed. From my previous experience I would always try to reduce cross cluster calls, but I am not sure how we feel about them at Kubermatic. Who would be the best person to weigh in on this?

File: docs/proposals/improve-addons.md
Comment (SimonTheLeg) [reply docs/proposals/improve-addons.md]: @mfranczy changed to ApplicationRef

File: docs/proposals/improve-addons.md
Comment (SimonTheLeg) [reply docs/proposals/improve-addons.md]: I have clarified the paragraph

File: docs/proposals/improve-addons.md
Comment (shaase-ctrl) [reply docs/proposals/improve-addons.md]: Using the seeds for such features is the intention of how I would like to evolve KKP in general. We deliver additional features via the seeds in order to steer traffic, access etc. 

1. UI and API, it is as you reasoned, in the beginning more UI, later on more automation via API. For an MVP we could start with UI and go with API later (but it should be possible). The spread of the user base is quite large in the end.

2. Very good point for the cross cluster calls, I do not think that we optimized for this, but it is a good consideration. Overall this is a clash between control for KKP admin and flexibility for user cluster admin. My thoughts on this revolve around the possibility to use helm anyway in a user cluster for the user cluster admin, but helm could be delivered via the "applications" feature. Control and flexibility if you will. If I understood correctly what @vgramer  said, kkp admins could involuntarily delete apps that are in use, which could be quite unfortunate. In order to get to a result, I think we should go with the kkp admin approach you already have for "applications" and see if there is a huge demand for independently manage "applications" via the user cluster admin. Thanks @mfranczy @SimonTheLeg @vgramer and @kron4eg for your considerations.

Comment (mfranczy) [reply docs/proposals/improve-addons.md]: @SimonTheLeg 

> It would be tough for cluster admins to create invalid ApplicationInstallationCRs

We should have a proper validation (on the client and server side), it should not matter if this is UI or API.

>  because if someone is more proficient in k8s and knows how to use helm, kustomized, etc., they would not really need the KKP Application feature. => now

Initially when I thought about this application mechanism I didn't mean to do it because some users don't know helm, it was for sake of automation and improved user experience (without forcing people to create their own automation on top to deploy standard apps they require - that's important with massive cluster creation)

> Reduce the number of cross-cluster calls. We need to be aware of the fact that moving the ApplicationInstallationCRs into the user clusters will result in more cross-cluster calls. This is due to the fact that we need to keep ApplicationDefinitions in the seed cluster (otherwise cluster admins could just edit the definitions and make up their own Applications).

Correct me if I am wrong, but IMO you won't reduce them here because the Application controller has to watch resources on the seed and user anyway. If `ApplicationInstallation` is deployed on the seed you have to check deployed resources on the user cluster to update the CR status. Or if `ApplicationInstallation` is on the user you have to get `ApplicationDefinition` CR.

@vgramer

> have a client to the seed cluster in the ApplicationInstallationController to be able to retrieve to the corresponding ApplicationDefinition CR.

Application controller has to be aware of seed and user clusters. Even with the current design while keeping the `ApplicationInstallation` CR on the seed you have to make a call to the user cluster to create a resource there over helm, k8s client or whatever source type you support.

Since the controller would run on the seed you have to get a client to user cluster (similar way how the current addon mechanism works).

If you move `ApplicationInstallation` CR to user you could have a custom informer to catch `ApplicationDefinition` events and map them to `ApplicationInstallation` CR events.


After all that's just my opinion, if you still feel different about that and have majority in this go ahead and continue with your idea.

**Comment by SimonTheLeg** [comment]
@vgramer and I were thinking a bit about @mfranczy comments in regards to moving the ApplicationInstallationCRs into the user cluster. After a nice afternoon we made the following observations:

- @mfranczy is absolutely correct on the cross cluster calls :) Because the user-k8s-api still has to fulfil the requests (which requires calls between clusters, as user-k8s-api lives in the seed), we do not really not save cross-cluster-calls
- we cannot offload everything into the user-cluster, as the ApplicationInstallationController also needs to handle secrets to external ressources (like git credentials)

long story short: we made a small adjustment, which both moves the ApplicationInstallationCR into the user-cluster while keeping the secret handling in the seed. You can quickly see the diff by using this link [`b1c3fed` (#8276)](https://github.com/kubermatic/kubermatic/pull/8276/commits/b1c3fed747812f9abee907511373118a8ec85b86).

cc: @kron4eg @shaase-ctrl  who were also part of the discussion

**Comment by kron4eg** [comment]
> @vgramer and I were thinking a bit about @mfranczy comments in regards to moving the ApplicationInstallationCRs into the user cluster. After a nice afternoon we made the following observations:
> 
>     * @mfranczy is absolutely correct on the cross cluster calls :) Because the user-k8s-api still has to fulfil the requests (which requires calls between clusters, as user-k8s-api lives in the seed), we do not really not save cross-cluster-calls
> 
>     * we cannot offload everything into the user-cluster, as the ApplicationInstallationController also needs to handle secrets to external ressources (like git credentials)
> 
> 
> long story short: we made a small adjustment, which both moves the ApplicationInstallationCR into the user-cluster while keeping the secret handling in the seed. You can quickly see the diff by using this link [`b1c3fed` (#8276)](https://github.com/kubermatic/kubermatic/pull/8276/commits/b1c3fed747812f9abee907511373118a8ec85b86).
> 
> cc: @kron4eg @shaase-ctrl who were also part of the discussion

:+1:

**Review by mfranczy — APPROVED** [review]
thanks for adjusting
/lgtm
/approve

**Comment by SimonTheLeg** [comment]
all pendings resolved. files have been renamed. 

@vgramer handing over to you for the final approve to merge this.

Thanks everyone for your work! 🥇

**Pull request merged**
